### DISCUSSION TITLE
How to build graph dataset from timeseries data

### DISCUSSION THREAD

**Discussion by neuronphysics** [pyg-team/pytorch_geometric#6341] [Q&A]
Hi,

I am new in using `pytorch_geometric` library. I looked up on the web to find any example related to building graph from timeseries data. However, I didn't come across anything useful. My questions are

-  How I can make a custom dataset for timeseries data where I have batch size, data size and sequence length along three dimensions of both inputs and outputs? I want to build a fully connected graph with the number of nodes would be the data size, and the value of the tensor would be the attribute of the node.
- I am wondering whether there is any method to deal with different sequence lengths when one uses a graph neural network to predict the outputs of a timeseries?
I would appreciate it if someone could direct me to github repositories or implement examples related to my questions.

Thanks in advance.

**Comment by rusty1s** [comment]
1. You should fine some success when storing your node features as `[num_nodes, num_seq_len, num_features]`.
2. You would be responsible for padding different sequence lenghts on your own, e.g., via `torch.nn.functional.pad`.

There also exists the PyTorch Geometric Temporal library which might be ideal for your use-case.

**Reply by mcunningham1440** [reply]
How do you add a temporal dimension to edge_index and edge_attr?

**Comment by neuronphysics** [comment]
Thanks for the reply. I have another question: How can I build a dataset when the structure of the input and target graph are different?  
If I use `StaticGraphTemporalSignal` the features of the input seems only evolve over time. Is there any way to include different graph structure for the output?

**Reply by rusty1s** [reply]
Are you referring to PyTorch Geometric Temporal? In this case, it might be best to ask for advice there, since I am not the author of this package.

**Comment by AidaSadeghi** [comment]
if you try to build your very own dataloader for a temporal dataset you should use Data object  inside pytorch geometric
- you can watch these tutorials on youtube which is so helpful https://www.youtube.com/playlist?list=PLV8yxwGOxvvoNkzPfCx2i8an--Tkt7O8Z
- you also can use this tensorflow timeseries tutorial Tensorflow time-series tutorial: [https://www.tensorflow.org/tutorials/...]
- in torch geometric temporal > dataset you can find the source of each dataset objects that are there 
- in this repo & especifically the notebook part you can see preproccessing of a traffic dataset: https://github.com/benedekrozemberczki/pytorch_geometric_temporal/tree/master/notebooks
